lib/propolis/src/hw/qemu/pvpanic.rs

```diff
@@ -0,0 +1,107 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::{Arc, Mutex};
+
+use crate::common::*;
+use crate::pio::{PioBus, PioFn};
+
+/// Implements the QEMU [pvpanic device], which
+/// may be used by guests to notify the host when a kernel panic has occurred.
+///
+/// QEMU exposes the pvpanic virtual device as a device on the ISA bus (I/O port
+/// 0x505), a PCI device, and through ACPI. Currently, Propolis only implements
+/// the ISA bus pvpanic device, but the PCI device may be implemented in the
+/// future.
+///
+/// [pvpanic device]: https://www.qemu.org/docs/master/specs/pvpanic.html
+#[derive(Debug)]
+pub struct QemuPvpanic {
+    counts: Mutex<Counts>,
+    log: slog::Logger,
+}
+
+#[derive(Debug)]
+struct Counts {
+    /// Counts the number of guest kernel panics handled by the host.
+    host_handled: usize,
+    /// Counts the number of guest kernel panics handled by the guest.
+    guest_handled: usize,
+}
+
+/// Indicates that a guest panic has happened and should be processed by the
+/// host
+const HOST_HANDLED: u8 = 0b01;
+/// Indicates a guest panic has happened and will be handled by the guest; the
+/// host should record it or report it, but should not affect the execution of
+/// the guest.
+const GUEST_HANDLED: u8 = 0b10;
+
+#[usdt::provider(provider = "propolis")]
+mod probes {
+    fn pvpanic_pio_write(value: u8) {}
+}
+
+impl QemuPvpanic {
+    const IOPORT: u16 = 0x505;
+
+    pub fn create(log: slog::Logger) -> Arc<Self> {
+        Arc::new(Self {
+            counts: Mutex::new(Counts { host_handled: 0, guest_handled: 0 }),
+            log,
+        })
+    }
+
+    /// Attaches this pvpanic device to the provided [`PioBus`].
+    pub fn attach_pio(self: &Arc<Self>, pio: &PioBus) {
+        let piodev = self.clone();
+        let piofn = Arc::new(move |_port: u16, rwo: RWOp| piodev.pio_rw(rwo))
+            as Arc<PioFn>;
+        pio.register(Self::IOPORT, 1, piofn).unwrap();
+    }
+
+    pub fn host_handled_count(&self) -> usize {
+        self.counts.lock().unwrap().host_handled
+    }
+
+    pub fn guest_handled_count(&self) -> usize {
+        self.counts.lock().unwrap().guest_handled
+    }
```

Comment:
Worth defining a struct with both values to emit all at once?  Seems like most consumers will just want one big snapshot.